---
title: "Complete DNA methylation analysis using Illumina EPICv1 microarray data"
author: "Kennedy Zinn"
date: "2025-03-18"
output: html_document
---

```{r, echo = FALSE}
suppressPackageStartupMessages(library(minfi))
suppressPackageStartupMessages(library(limma))
suppressPackageStartupMessages(library(DMRcate))
suppressPackageStartupMessages(library(maxprobes))
suppressPackageStartupMessages(library(readxl))
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(missMethyl))
suppressPackageStartupMessages(library(MethylToSNP))
```

# Smooth Muscle Cell (SMC) Comparisons

## Load Data

```{r, echo = FALSE}
rgSet <- read.metharray.exp("/path/to/idat/folder", force = T)
```

load sample sheet

```{r, echo = FALSE}
sample_sheet <- read_excel("/path/to/excel/sample/sheet")
```

## Quality check

```{r, echo = FALSE}
suppressMessages(qcReport(rgSet, pdf = "/path/to/new/qcReport.pdf", sampNames = sample_sheet$`Sample Name`, sampGroups = sample_sheet$Status))
```

## Pre-Process
```{r}
source("process_methyl_microarray.R")
gmSet <- process_methyl_microarray(rgSet)
```

Beta values, which range between [0,1], represent the proportion of methylated DNA copies at a specific CpG site compared to all of the DNA copies at that site.  Beta values for the ith CpG site are mathematically defined as,

Beta(i) = max(yi(methyl),0) / [max(yi(unmethyl), 0) + max(yi(methyl), 0) + alpha],

where yi methyl and unmethyl are the intensities measured by the ith methylated and unmethylated probes, respectively. 0 ensures that negative signals that take place after background adjustment are not considered. Alpha, a constant (usually 100), serves as an offset to regularize beta when both methylated and unmethylated instensities are low. Beta values are intuitive, and hence useful for visual presentations of the reads. (Du et al., 2010)

```{r, echo = FALSE}
# Compute beta values
beta_values <- getBeta(gmSet)
colnames(beta_values) <- sample_sheet$`Sample Name`
```

M values are log transformed beta values. Negative M values indicate higher levels of unmethylated DNA copies than methylated DNA copies. Postive M values indicate higher levels of methylated DNA copies than unmethylated. M values close to zero indicate similar intensities between methylated and unmethylated probes. M values are mathematically expressed as, 

M(i) = log2([max(yi(methyl))+alpha]/[max(yi(unmethyl))+alpha])

where alpha is a constant to offset dramatic changes resulting from small estimation errors. M values can be expressed in terms of beta values,

M(i) = log2(Beta(i)/[1 - Beta(i)])

and therefore, beta values can be expressed in terms of M values,

Beta(i) = 2^M(i) / [2^M(i) + 1]

M values are more statistically sound than beta values.

```{r, echo = FALSE}
Mval <- getM(gmSet)
dim(Mval)
```

Use the PCA plot to identify possible sources of variation other than the factor of interest. In my case, samples clustered by Sentrix ID, suggesting a batch effect. Identifying such sources of variation can be included in the model during differential analysis.
```{r, echo = FALSE}
# PCA plot of M values
par(mfrow=c(1,1))
plotMDS(Mval, labels=sample_sheet$`Sample Name`, col=as.integer(factor(sample_sheet$Status)))
legend("top",legend=c("Control","Patient"),pch=16,cex=.5,col=1:2)
```

## Differential Methylation Analysis

Design matrix

```{r, echo = FALSE}
batch <- factor(ifelse(sample_sheet$`Sentrix ID` == 20408156, "batch.effect", "no.batch"))
disease <- factor(sample_sheet$Status)
design <- model.matrix(~1 + disease + batch, data = sample_sheet)
colnames(design) <- c(levels(disease),levels(batch)[-1])
design
```

## Removing Unwanted Variation with SVA package

Fit model using M values as M values are more statistically valid, beta is more intuitive (Du et al., 2010)

```{r}
# create full and null model
full <- design
null <- model.matrix(~1 + batch, data = sample_sheet)

# applying sva function
library(sva)
n.sv <- num.sv(Mval,full,method = "leek") # estimate number of latent factors
n.sv
# estimate the surrogate variables (variables which are intended to represent the unknown variation)
svobj <- sva(Mval, full, null, n.sv = n.sv)

# adjust surrogate variable using f-test p values of each cpg site
f_p_val <- f.pvalue(Mval,full,null)
q_val <- p.adjust(f_p_val, method = "BH")

fullSv <- cbind(full, svobj$sv)
nullSv <- cbind(null, svobj$sv)
p_val_sv <- f.pvalue(Mval, fullSv, nullSv)
q_val_sv <- p.adjust(p_val_sv, method = "BH")

# run model while adjusting for surrogate variables
fit <- lmFit(Mval, fullSv)
colnames(fullSv) <- make.names(colnames(fullSv))
contrast <- makeContrasts(control - patient, levels = fullSv)
fitContrast <- contrasts.fit(fit, contrast)
fit <- eBayes(fitContrast)
top <- topTable(fit, adjust = "BH", number = "Inf")
head(top)
```

Visualize results using beta values

```{r, echo = FALSE}
cpgs <- rownames(top)
par(mfrow = c(2,2), cex = 0.5)
for(i in 1:4){
stripchart(beta_values[rownames(beta_values) %in% cpgs[i],]~design[,2],method="jitter",
group.names=c("Control","Patient"), pch = 16,cex=1.5,col=c(4,2), ylab="Beta values",
vertical=TRUE,cex.axis=1.5,cex.lab=1.5)
title(cpgs[i],cex.main=1.5)
}
```

## Differential Variability Analysis

fit model testing variances between controls and patients

```{r, echo = FALSE}
fitvar <- varFit(Mval, design = design, coef = 1:3)
summary(decideTests(fitvar))
topVar <- topVar(fitvar, coef=3)
topVar
```

Visualize top 4 most differentially variable CpGs

```{r, echo = FALSE}
cpgsDV <- rownames(topVar)
par(mfrow=c(2,2), cex = 0.5)
for(i in 1:4){
stripchart(beta_values[rownames(beta_values)==cpgsDV[i],]~design[,2], method="jitter",
group.names=c("Control","Patient"),pch=16,cex=1.5,col=c(4,2),ylab="Beta values",
vertical=TRUE,cex.axis=1.5,cex.lab=1.5)
title(cpgsDV[i],cex.main=1.5)
}
```

Identify the most variable patient sample

```{r, echo = FALSE}
par(mfrow=c(2,2), cex = 0.5)
for(i in 1:4){
stripchart(beta_values[rownames(beta_values)==cpgsDV[i],]~colnames(beta_values), method="jitter",pch=16,cex=1.5,ylab="Beta values",
vertical=TRUE,cex.axis=1.5,cex.lab=1.5)
title(cpgsDV[i],cex.main=1.5)
}
```

## Gene Ontology Analysis

### CpG Analysis

```{r, echo = F}
# Observe differentially methylated CpGs using 1% FDR cut-off
table(top$adj.P.Val < 0.01)
```

More than 10k significant cpg sites is too much to yield biologically relevant results. We can trim the data by introducing a cutoff of the change in beta values.

```{r}
# make sure that order of beta values matches order after analysis
beta <- getBeta(gmSet)
beta <- beta[match(rownames(top),rownames(beta)),]
beta_ctrl <- rowMeans(beta[,disease=="control"])
beta_pt <- rowMeans(beta[,disease=="patient"])
Delta_beta <- beta_ctrl - beta_pt
sigDM <- top$adj.P.Val < 0.01 & abs(Delta_beta)>0.25
table(sigDM)
```

```{r}
topCpGs <- beta[sigDM == TRUE,]
sigCpGs <- rownames(topCpGs)
library(IlluminaHumanMethylationEPICanno.ilm10b5.hg38)
data(IlluminaHumanMethylationEPICanno.ilm10b5.hg38)
anno <- getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b5.hg38)
check <- getMappedEntrezIDs(sig.cpg = sigCpGs, array.type = "EPIC", anno = anno)
```

```{r}
# Check number of genes to which CpGs are annotated
length(check$sig.eg)
```

```{r}
gst.go <- gometh(sig.cpg=sigCpGs, all.cpg=rownames(top), collection="GO", array.type = "EPIC",
              plot.bias=TRUE, anno = anno, sig.genes = T) # Include "all.cpg" option if filtering CpGs prior to testing
```

```{r}
topGSA(gst.go[, -ncol(gst.go)])
```

We can also test according to a specified gene set using "gsameth"

### Region level analysis

```{r}
source("/Users/kennedyzinn/Desktop/OneDrive - UTHealth Houston/GRA/dna methyl/region_analysis.R")
results.ranges <- region_analysis("control - patient", "0.25", "0.01")
```

```{r}
gst.region.go <- goregion(results.ranges, all.cpg=rownames(top), 
                       collection= "GO", array.type="EPIC", plot.bias = T, anno = anno, sig.genes = T)
topGSA(gst.region.go[, -ncol(gst.region.go)])
```

We can also test according to a specified gene set using "gsaregion"

## KEGG Analysis

### CpG Analysis

```{r}
gst.kegg <- gometh(sig.cpg=sigCpGs, all.cpg=rownames(top), collection="KEGG", array.type = "EPIC", anno = anno, plot.bias=TRUE, sig.genes = T) # Include "all.cpg" option if filtering CpGs prior to testing
```

```{r}
topGSA(gst.kegg[, -ncol(gst.kegg)])
# We can also test according to a specified gene set using "gsameth"
```

### Region level analysis

```{r}
gst.region.kegg <- goregion(results.ranges, all.cpg=rownames(top), 
                       collection= "KEGG", array.type="EPIC", plot.bias = T, anno = anno, sig.genes = T)
topGSA(gst.region.kegg[, -ncol(gst.region.kegg)]) 
```

# Identify the direction of methylation in significant gene sets
```{r}
sig.GeneSets <- unique(c(gst.go$SigGenesInSet[gst.go$FDR < 0.05], 
                         gst.region.go$SigGenesInSet[gst.region.go$FDR < 0.05], 
                         gst.kegg$SigGenesInSet[gst.kegg$FDR <0.05,], 
                         gst.region.kegg$SigGenesInSet[gst.region.kegg$FDR < 0.05]))
# Split the gene sets into individual genes
split_genes <- function(gene_set) {
 unlist(strsplit(gene_set, split = "[,]"))
}

# Apply the splitting function to all significant gene sets
all_genes <- unique(unlist(lapply(sig.GeneSets, split_genes)))
```

```{r}

# Annotate probes
library(tibble)
top <- rownames_to_column(top, var = "ID")
anno <- as.data.frame(anno)
anno <- rownames_to_column(anno, var = "ID")
top <- merge(top, anno, by = "ID")
```

```{r}
# Split UCSC_RefGene_Name into individual genes
library(tidyr)
top <- top %>%
  mutate(UCSC_RefGene_Name = strsplit(as.character(UCSC_RefGene_Name), split = "[,;]")) %>%
  unnest(UCSC_RefGene_Name)

# Filter for significant gene sets
filteredResults <- top[top$UCSC_RefGene_Name %in% all_genes, ]

```

```{r}

# Summarize methylation status for each significant gene set
methylationSummary <- sapply(all_genes, function(gene) {
 probes <- filteredResults[filteredResults$UCSC_RefGene_Name == gene, ]
 avgLogFC <- mean(probes$logFC)
 return(avgLogFC)
})
```

```{r}
# Determine methylation status
methylationStatus <- ifelse(methylationSummary > 0, "Hypermethylated", "Hypomethylated")

# Combine results into a data frame
results <- data.frame(GeneSet = all_genes, MethylationStatus = methylationStatus)

print(results)
```
