---
title: "IPSC analysis"
author: "Kennedy Zinn"
date: "2025-02-07"
output: html_document
---

```{r}
suppressPackageStartupMessages(library(minfi))
suppressPackageStartupMessages(library(limma))
suppressPackageStartupMessages(library(DMRcate))
suppressPackageStartupMessages(library(maxprobes))
suppressPackageStartupMessages(library(readxl))
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(missMethyl))
suppressPackageStartupMessages(library(MethylToSNP))
```

# Induced Pluripotent Stem Cell (IPSC) Comparisons

## Load Data

```{r}
# load the idat files
rgSet <- read.metharray.exp("/Users/kennedyzinn/Desktop/OneDrive - UTHealth Houston/GRA/dna methyl/IPSC")
```

```{r}
# load sample sheet
sample_sheet <- read_excel("/Users/kennedyzinn/Desktop/OneDrive - UTHealth Houston/GRA/dna methyl/IPSC/SampleSheet_IPSC.xlsx")
```

## Quality check

```{r}
# Generate QC report
qcReport(rgSet, pdf = "/Users/kennedyzinn/Downloads/IPSCqcReport.pdf", sampNames = sample_sheet$`Sample Name`, sampGroups = sample_sheet$Status)
```

```{r}
source("/Users/kennedyzinn/Desktop/OneDrive - UTHealth Houston/GRA/dna methyl/process_methyl_microarray.R")
gmSet <- process_methyl_microarray(rgSet)
```

Beta values, which range between [0,1], represent the proportion of methylated DNA copies at a specific CpG site compared to all of the DNA copies at that site. Beta values for the ith CpG site are mathematically defined as,

Beta(i) = max(yi(methyl),0) / [max(yi(unmethyl), 0) + max(yi(methyl), 0) + alpha],

where yi methyl and unmethyl are the intensities measured by the ith methylated and unmethylated probes, respectively. 0 ensures that negative signals that take place after background adjustment are not considered. Alpha, a constant (usually 100), serves as an offset to regularize beta when both methylated and unmethylated instensities are low. Beta values are intuitive, and hence useful for visual presentations of the reads. (Du et al., 2010)

```{r, echo = FALSE}
# Compute beta values
beta_values <- getBeta(gmSet)
colnames(beta_values) <- sample_sheet$`Sample Name`
```

M values are log transformed beta values. Negative M values indicate higher levels of unmethylated DNA copies than methylated DNA copies. Postive M values indicate higher levels of methylated DNA copies than unmethylated. M values close to zero indicate similar intensities between methylated and unmethylated probes. M values are mathematically expressed as,

M(i) = log2([max(yi(methyl))+alpha]/[max(yi(unmethyl))+alpha])

where alpha is a constant to offset dramatic changes resulting from small estimation errors. M values can be expressed in terms of beta values,

M(i) = log2(Beta(i)/[1 - Beta(i)])

and therefore, beta values can be expressed in terms of M values,

Beta(i) = 2\^M(i) / [2\^M(i) + 1]

M values are more statistically sound than beta values.

```{r, echo = FALSE}
# Compute M values
Mval <- getM(gmSet)
dim(Mval)
```

Use the PCA plot to infer what factors are driving variation. Here the difference in 18APCK from the rest of the samples seems to drive 20% of total variation (bad). Unsure what accounts for the first PC. Maybe by who processed the samples?

```{r}
# PCA of M values
par(mfrow=c(1,1), cex =.5)
plotMDS(Mval, labels=sample_sheet$`Sample Name`, col=as.integer(factor(sample_sheet$Status)))
legend("top",legend=c("Control","Patient"),pch=16,col=1:2)
```

## Differential Methylation Analysis

```{r, echo = FALSE}
# Design matrix
disease <- factor(sample_sheet$Status)
design <- model.matrix(~1 + disease, data = sample_sheet)
colnames(design) <- c(levels(disease))
design
```

Fit model using M values as M values are more statistically valid, beta is more intuitive (Du et al., 2010)

```{r}
fit.reduced <- lmFit(Mval, design)
fit <- eBayes(fit)
```

```{r, echo = FALSE}
# Model Results
summary(decideTests(fit))
top<-topTable(fit,coef=1, number = Inf)
head(top)
```

```{r}
# Visualize results using beta values
cpgs <- rownames(top)
par(mfrow = c(2,2), cex = 0.5)
for(i in 1:4){
stripchart(beta_values[rownames(beta_values) %in% cpgs[i],]~design[,2],method="jitter",
group.names=c("Control","Patient"), pch = 16,cex=1.5,col=c(4,2), ylab="Beta values",
vertical=TRUE,cex.axis=1.5,cex.lab=1.5)
title(cpgs[i],cex.main=1.5)
}
```

## Differential Methylation Analysis while adjusting for unwanted variation

```{r}
grp <- factor(sample_sheet$`Status`, labels=c(0,1))
INCs <- getINCs(rgSet)
head(INCs)
Mc <- rbind(Mval,INCs)
ctl1 <- rownames(Mc) %in% rownames(INCs)
rfit1 <- RUVfit(Y = Mc, X = grp, ctl = ctl1)
rfit2 <- RUVadj(Y = Mc, fit = rfit1)
top1 <- topRUV(rfit2, num=Inf, p.BH = 1)
head(top1)
```

```{r}
ctl2 <- rownames(Mval) %in% rownames(top1[top1$p.BH_X1.1 > 0.5,])
table(ctl2)
```

```{r}
rfit3 <- RUVfit(Y = Mval, X = grp, ctl = ctl2) # Stage 2 analysis
rfit4 <- RUVadj(Y = Mval, fit = rfit3)
topRUV(rfit4)
```

How well did RUV perform? Due to the decrease in number of discovered genes, the unwanted variation may be correlated with moya-moya (Gagnon-Bartsch & Speed, 2012). Which samples were processed by Dr. K, cambridge, etc? this might be key (or, it might already be accounted for in batch factor)
```{r}
par(mfrow = c(1,2))
hist(top$adj.P.Val)
top2 <- topRUV(rfit4, num = Inf)
hist(top2$p.BH_X1.1)
```


## Differential Variability Analysis

```{r}
# fit model testing variances between controls and patients
fitvar <- varFit(Mval, design = design, coef = c(1,2))
summary(decideTests(fitvar))
topVar <- topVar(fitvar, coef=2)
topVar
```

```{r}
# Visualize top 4 most differentially variable CpGs
cpgsDV <- rownames(topVar)
par(mfrow=c(2,2), cex = 0.5)
for(i in 1:4){
stripchart(beta_values[rownames(beta_values)==cpgsDV[i],]~design[,2],method="jitter",
group.names=c("Control","Patient"),pch=16,cex=1.5,col=c(4,2),ylab="Beta values",
vertical=TRUE,cex.axis=1.5,cex.lab=1.5)
title(cpgsDV[i],cex.main=1.5)
}
```

```{r}
# Identify the most variable patient sample
par(mfrow=c(2,2), cex = 0.5)
for(i in 1:4){
stripchart(beta_values[rownames(beta_values)==cpgsDV[i],]~colnames(beta_values), method="jitter",pch=16,cex=1.5,ylab="Beta values",
vertical=TRUE,cex.axis=1.5,cex.lab=1.5)
title(cpgsDV[i],cex.main=1.5)
}
```


## Gene Ontology Analysis

### CpG Analysis

```{r, echo = F}
# Observe differentially methylated CpGs using 1% FDR cut-off
table(top$adj.P.Val < 0.01)
```

655k significant cpg sites is too much to yield biologically relevant results. We can trim the data by introducing a cutoff of the change in beta values.

```{r}
# make sure that order of beta values matches order after analysis
beta <- getBeta(gmSet)
beta <- beta[match(rownames(top),rownames(beta)),]
beta_ctrl <- rowMeans(beta[,disease=="control"])
beta_pt <- rowMeans(beta[,disease=="patient"])
Delta_beta <- beta_ctrl - beta_pt
sigDM <- top$adj.P.Val < 0.01 & abs(Delta_beta) > 0.05
table(sigDM)
```

```{r}
topCpGs <- beta[sigDM == TRUE,]
sigCpGs <- rownames(topCpGs)
check <- getMappedEntrezIDs(sig.cpg = sigCpGs, array.type = "EPIC")
```

```{r}
# Check number of genes to which CpGs are annotated
length(check$sig.eg)
```

```{r}
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
gst <- gometh(sig.cpg=sigCpGs, all.cpg=rownames(top), collection="GO", array.type = "EPIC",
              plot.bias=TRUE) # Include "all.cpg" option if filtering CpGs prior to testing
```

```{r}
topGSA(gst)
```

We can also test according to a specified gene set using "gsameth"

### Region level analysis

```{r}
source("/Users/kennedyzinn/Desktop/OneDrive - UTHealth Houston/GRA/dna methyl/region_analysis.R")
region_analysis("control - patient", "GO", "0.1", "0.01")
```

We can also test according to a specified gene set using "gsaregion"

## KEGG Analysis

### CpG Analysis

```{r}
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
gst <- gometh(sig.cpg=sigCpGs, all.cpg=rownames(top), collection="KEGG", array.type = "EPIC",
              plot.bias=TRUE) # Include "all.cpg" option if filtering CpGs prior to testing
```

```{r}
topGSA(gst)
```

We can also test according to a specified gene set using "gsameth"

### Region level analysis

```{r}
region_analysis("control - patient", "KEGG", "0.1", "0.01")
```
